cmd/lotus-miner/is_stop_safe.go

```diff
@@ -0,0 +1,318 @@
+	}
+	return errs, &maddr, cd
+}
+
+func GetMarketErrors(cctx *cli.Context) []error {
+	api, closer, err := lcli.GetMarketsAPI(cctx)
+	if err != nil {
+		return []error{err}
+	}
+	defer closer()
+
+	ctx := lcli.DaemonContext(cctx)
+
+	onlineStorageDealsOk, err := api.DealsConsiderOnlineStorageDeals(ctx)
+	if err != nil {
+		return []error{err}
+	}
+
+	onlineRetrievalDealsOk, err := api.DealsConsiderOnlineRetrievalDeals(ctx)
+	if err != nil {
+		return []error{err}
+	}
+
+	errs := []error{}
+	if onlineStorageDealsOk {
+		errs = append(errs, xerrors.Errorf("Online storage deals are being accepted"))
+	}
+	if onlineRetrievalDealsOk {
+		errs = append(errs, xerrors.Errorf("Online retrieval deals are being accepted"))
+	}
+
+	errs = append(errs, GetStorageDealErrors(cctx, api)...)
+	errs = append(errs, GetRetrievalDealErrors(cctx, api)...)
+	errs = append(errs, GetDataTransferErrors(cctx, api)...)
+
+	return errs
+}
+
+func GetStorageDealErrors(cctx *cli.Context, api api.StorageMiner) []error {
+	deals, err := api.MarketListIncompleteDeals(lcli.DaemonContext(cctx))
+	if err != nil {
+		return []error{err}
+	}
+
+	errs := []error{}
+	for _, deal := range deals {
+		if _, ok := SafeStorageDealStates[deal.State]; !ok {
+			errs = append(errs, xerrors.Errorf("Storage deal is not in a safe state: %d = %s", deal.DealID, storagemarket.DealStates[deal.State]))
+		}
+	}
+	return errs
+}
+
+func GetRetrievalDealErrors(cctx *cli.Context, api api.StorageMiner) []error {
+	deals, err := api.MarketListRetrievalDeals(lcli.DaemonContext(cctx))
+	if err != nil {
+		return []error{err}
+	}
+
+	errs := []error{}
+	for _, deal := range deals {
+		if _, ok := SafeRetrievalDealStatuses[deal.Status]; !ok {
+			errs = append(errs, xerrors.Errorf("Retrieval deal is not in a safe state: %d = %s", deal.ID, retrievalmarket.DealStatuses[deal.Status]))
+		}
+	}
+	return errs
+}
+
+func GetDataTransferErrors(cctx *cli.Context, api api.StorageMiner) []error {
+	ctx := lcli.ReqContext(cctx)
+
+	channels, err := api.MarketListDataTransfers(ctx)
+	if err != nil {
+		return []error{err}
+	}
+
+	errs := []error{}
+	for _, channel := range channels {
+		if _, ok := SafeDataTransferStatuses[channel.Status]; !ok {
+			errs = append(errs, xerrors.Errorf("Data transfer is not in a safe state: %d = %s", channel.TransferID, datatransfer.Statuses[channel.Status]))
+		}
+	}
+	return errs
+}
+
+var SafeStorageDealStates = map[storagemarket.StorageDealStatus]struct{}{
+	storagemarket.StorageDealUnknown:          {},
+	storagemarket.StorageDealProposalNotFound: {},
+	storagemarket.StorageDealProposalRejected: {},
+	// storagemarket.StorageDealProposalAccepted:             {},
+	// storagemarket.StorageDealAcceptWait:                   {},
+	// storagemarket.StorageDealStartDataTransfer:            {},
+	// storagemarket.StorageDealStaged:                       {},
+	// storagemarket.StorageDealAwaitingPreCommit:            {},
+	// storagemarket.StorageDealSealing:                      {},
```

Comment:
It should be ok to stop when deals are in the sealing pipeline, sealing should be able to restart cleanly